File: lib/services/progress_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';

class ProgressService {
static final _db = FirebaseFirestore.instance;
static final _auth = FirebaseAuth.instance;

static String _uidOrThrow() {
final uid = _auth.currentUser?.uid;
if (uid == null) throw Exception("No user signed in.");
return uid;
}

static DocumentReference<Map<String, dynamic>> _userDoc() {
final uid = _uidOrThrow();
return _db.collection('users').doc(uid);
}

static CollectionReference<Map<String, dynamic>> _daysCol() {
return _userDoc().collection('days');
}

static String dateKeyFor(DateTime dt) {
final y = dt.year.toString().padLeft(4, '0');
final m = dt.month.toString().padLeft(2, '0');
final d = dt.day.toString().padLeft(2, '0');
return '$y-$m-$d';
}

static String todayKey() => dateKeyFor(DateTime.now());

static String yesterdayKey() =>
dateKeyFor(DateTime.now().subtract(const Duration(days: 1)));

static int _nowMs() => DateTime.now().millisecondsSinceEpoch;

static int? _asInt(dynamic raw) {
if (raw is int) return raw;
if (raw is num) return raw.toInt();
return null;
}

/// Ensures /users/{uid} exists and required retention fields exist.
static Future<void> ensureUserDoc() async {
final ref = _userDoc();
final snap = await ref.get();

final nowMs = _nowMs();

final base = <String, dynamic>{
'points': 0,
'currentStreak': 0,
'longestStreak': 0,
'lastCallAtMs': null,
'lastCompletedCallAtMs': null,
'lastActiveAtMs': nowMs,
'createdAt': FieldValue.serverTimestamp(),
'updatedAt': FieldValue.serverTimestamp(),
};

if (!snap.exists) {
await ref.set(base, SetOptions(merge: true));
return;
}

final data = snap.data() ?? <String, dynamic>{};

await ref.set(
{
'updatedAt': FieldValue.serverTimestamp(),
if (!data.containsKey('points')) 'points': 0,
if (!data.containsKey('currentStreak')) 'currentStreak': 0,
if (!data.containsKey('longestStreak')) 'longestStreak': 0,
if (!data.containsKey('lastCallAtMs')) 'lastCallAtMs': null,
if (!data.containsKey('lastCompletedCallAtMs'))
'lastCompletedCallAtMs': null,
if (!data.containsKey('lastActiveAtMs')) 'lastActiveAtMs': nowMs,
},
SetOptions(merge: true),
);
}

/// Reads /users/{uid}
static Future<Map<String, dynamic>> getUserProgress() async {
final snap = await _userDoc().get();
return snap.data() ?? <String, dynamic>{};
}

/// Writes /users/{uid} (merge)
static Future<void> setUserProgress(Map<String, dynamic> data) async {
await _userDoc().set(
{
...data,
'updatedAt': FieldValue.serverTimestamp(),
},
SetOptions(merge: true),
);
}

/// App-open / screen-open activity heartbeat.
static Future<void> touchActiveNow() async {
final nowMs = _nowMs();
await _userDoc().set(
{
'lastActiveAtMs': nowMs,
'updatedAt': FieldValue.serverTimestamp(),
},
SetOptions(merge: true),
);
}

/// Call-complete truth for retention logic.
static Future<void> markCallCompletedNow() async {
final nowMs = _nowMs();
await _userDoc().set(
{
'lastActiveAtMs': nowMs,
'lastCallAtMs': nowMs,
'lastCompletedCallAtMs': nowMs,
'updatedAt': FieldValue.serverTimestamp(),
},
SetOptions(merge: true),
);
}

static Future<int?> getLastActiveAtMs() async {
final data = await getUserProgress();
return _asInt(data['lastActiveAtMs']);
}

static Future<int?> getLastCompletedCallAtMs() async {
final data = await getUserProgress();
return _asInt(data['lastCompletedCallAtMs']);
}

/// Used by matching eligibility.
static Future<bool> isRecentlyActive({
int maxInactiveDays = 3,
}) async {
final lastActiveAtMs = await getLastActiveAtMs();
if (lastActiveAtMs == null) return false;

final nowMs = _nowMs();
final cutoffMs = Duration(days: maxInactiveDays).inMilliseconds;

return (nowMs - lastActiveAtMs) <= cutoffMs;
}

/// Reads /users/{uid}/days/{dateKey}
static Future<Map<String, dynamic>?> getDay(String dateKey) async {
final snap = await _daysCol().doc(dateKey).get();
if (!snap.exists) return null;
return snap.data();
}

static Future<bool> didCompleteDay(String dateKey) async {
final day = await getDay(dateKey);
if (day == null) return false;
return (day['callCompleted'] ?? false) == true;
}

static Future<bool> wasDayMissed(String dateKey) async {
final day = await getDay(dateKey);
if (day == null) return false;

final hiddenName = (day['hiddenName'] ?? '').toString().trim();
final completed = (day['callCompleted'] ?? false) == true;

if (hiddenName.isEmpty) return false;
if (hiddenName == 'Add more friends') return false;
if (completed) return false;

return true;
}

static Future<bool> wasYesterdayMissed() async {
return wasDayMissed(yesterdayKey());
}

/// Writes /users/{uid}/days/{dateKey} (merge)
static Future<void> setDay(String dateKey, Map<String, dynamic> data) async {
final merged = <String, dynamic>{
...data,
'dateKey': dateKey,
'updatedAt': FieldValue.serverTimestamp(),
};

await _daysCol().doc(dateKey).set(merged, SetOptions(merge: true));
}
}

File: ios-native/Services/ProgressService.swift
import Foundation
import FirebaseAuth
import FirebaseFirestore

public final class ProgressService {
    private let db: Firestore
    private let auth: Auth

    public init(
        firestore: Firestore = Firestore.firestore(),
        auth: Auth = Auth.auth()
    ) {
        self.db = firestore
        self.auth = auth
    }

    public static func dateKey(for date: Date) -> String {
        let components = Calendar.current.dateComponents([.year, .month, .day], from: date)
        let year = String(format: "%04d", components.year ?? 0)
        let month = String(format: "%02d", components.month ?? 0)
        let day = String(format: "%02d", components.day ?? 0)
        return "\(year)-\(month)-\(day)"
    }

    public func ensureUserDoc() async throws {
        let reference = try userDoc()
        let snapshot = try await reference.getDocument()
        let nowMs = Self.nowMs()

        let base: [String: Any] = [
            "points": 0,
            "currentStreak": 0,
            "longestStreak": 0,
            "lastCallAtMs": NSNull(),
            "lastCompletedCallAtMs": NSNull(),
            "lastActiveAtMs": nowMs,
            "createdAt": FieldValue.serverTimestamp(),
            "updatedAt": FieldValue.serverTimestamp(),
        ]

        if !snapshot.exists {
            try await reference.setData(base, merge: true)
            return
        }

        let data = snapshot.data() ?? [:]
        var payload: [String: Any] = [
            "updatedAt": FieldValue.serverTimestamp(),
        ]

        if data["points"] == nil {
            payload["points"] = 0
        }
        if data["currentStreak"] == nil {
            payload["currentStreak"] = 0
        }
        if data["longestStreak"] == nil {
            payload["longestStreak"] = 0
        }
        if data["lastCallAtMs"] == nil {
            payload["lastCallAtMs"] = NSNull()
        }
        if data["lastCompletedCallAtMs"] == nil {
            payload["lastCompletedCallAtMs"] = NSNull()
        }
        if data["lastActiveAtMs"] == nil {
            payload["lastActiveAtMs"] = nowMs
        }

        try await reference.setData(payload, merge: true)
    }

    public func getUserProgress() async throws -> [String: Any] {
        let snapshot = try await userDoc().getDocument()
        return snapshot.data() ?? [:]
    }

    public func setUserProgress(_ data: [String: Any]) async throws {
        var payload = data
        payload["updatedAt"] = FieldValue.serverTimestamp()
        try await userDoc().setData(payload, merge: true)
    }

    public func touchActiveNow() async throws {
        try await userDoc().setData(
            [
                "lastActiveAtMs": Self.nowMs(),
                "updatedAt": FieldValue.serverTimestamp(),
            ],
            merge: true
        )
    }

    public func markCallCompletedNow() async throws {
        let nowMs = Self.nowMs()
        try await userDoc().setData(
            [
                "lastActiveAtMs": nowMs,
                "lastCallAtMs": nowMs,
                "lastCompletedCallAtMs": nowMs,
                "updatedAt": FieldValue.serverTimestamp(),
            ],
            merge: true
        )
    }

    public func isRecentlyActive(maxInactiveDays: Int = 3) async throws -> Bool {
        guard let lastActiveAtMs = try await getLastActiveAtMs() else {
            return false
        }

        let cutoffMs = maxInactiveDays * 24 * 60 * 60 * 1000
        return (Self.nowMs() - lastActiveAtMs) <= cutoffMs
    }

    public func getDay(_ dateKey: String) async throws -> [String: Any]? {
        let snapshot = try await daysCollection().document(dateKey).getDocument()
        guard snapshot.exists else {
            return nil
        }
        return snapshot.data()
    }

    public func wasDayMissed(_ dateKey: String) async throws -> Bool {
        guard let day = try await getDay(dateKey) else {
            return false
        }

        let hiddenName = ((day["hiddenName"] ?? "") as AnyObject)
            .description?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let completed = (day["callCompleted"] as? Bool) == true

        if hiddenName.isEmpty {
            return false
        }
        if hiddenName == "Add more friends" {
            return false
        }
        if completed {
            return false
        }

        return true
    }

    public func wasYesterdayMissed() async throws -> Bool {
        try await wasDayMissed(Self.yesterdayKey())
    }

    public func setDay(_ dateKey: String, data: [String: Any]) async throws {
        var merged = data
        merged["dateKey"] = dateKey
        merged["updatedAt"] = FieldValue.serverTimestamp()
        try await daysCollection().document(dateKey).setData(merged, merge: true)
    }

    private func uidOrThrow() throws -> String {
        guard let uid = auth.currentUser?.uid else {
            throw ProgressServiceError.noUserSignedIn
        }
        return uid
    }

    private func userDoc() throws -> DocumentReference {
        db.collection("users").document(try uidOrThrow())
    }

    private func daysCollection() throws -> CollectionReference {
        try userDoc().collection("days")
    }

    private func getLastActiveAtMs() async throws -> Int? {
        let data = try await getUserProgress()
        return Self.asInt(data["lastActiveAtMs"])
    }

    private func getLastCompletedCallAtMs() async throws -> Int? {
        let data = try await getUserProgress()
        return Self.asInt(data["lastCompletedCallAtMs"])
    }

    private static func todayKey() -> String {
        dateKey(for: Date())
    }

    private static func yesterdayKey() -> String {
        dateKey(for: Date().addingTimeInterval(-86_400))
    }

    private static func nowMs() -> Int {
        Int(Date().timeIntervalSince1970 * 1000)
    }

    private static func asInt(_ raw: Any?) -> Int? {
        if let value = raw as? Int {
            return value
        }
        if let value = raw as? NSNumber {
            return value.intValue
        }
        return nil
    }
}

public enum ProgressServiceError: Error {
    case noUserSignedIn
}